git-branchless/src/commands/smartlog.rs

```diff
@@ -566,13 +758,13 @@ pub fn smartlog(
         &references_snapshot,
     )?;
 
-    let commits = match resolve_commits(
-        effects,
-        &repo,
-        &mut dag,
-        &[revset.clone()],
-        resolve_revset_options,
-    ) {
+    let (revset, smartlog_variant) = match revset {
```

Comment:
An alternative could be to define `SmartlogVariant` as

```
enum SmartlogVariant {
    Connected,
    Sparse(Revset),
}
```

and then have this field be of type `SmartlogVariant` instead. Then it will have a sensible default that will be evaluated later. You could define a `SmartlogVariant::revset()`/`SmartlogVariant::get_revset()` function to get the associated revset or default, or just `match` on it each time.

Reply:
This went well. I hope you meant to suggest making the `revset` argument on the CLI into a `SmartlogVariant` and then giving clap the tools to turn a string from the CLI into a `SmartlogVariant::Sparse(Revset(...))`, because that's what I did.

One slightly odd thing about how I did this is that the contained revset is only needed/used once: in `fn smartlog`, to unpack the revset that needs to be evaluated and rendered. After that, we continue using the `variant`, but because the revset is no longer needed, all of the matching happens on `Connected` or `Sparse(_)`. Maybe this is fine, but it felt a 🤔 to consistently ignore that field. Maybe I didn't understand what you meant, though; I'm happy to make an update if so!